File: src/lib/server/apkg/ir.ts
/**
 * The intermediate representation between an Anki package and our schema.
 *
 * `import: apkg -> IR -> db`, `export: db -> IR -> apkg` (CLAUDE.md §4). Nothing in here is a
 * Drizzle row, and nothing in here still carries a format quirk: this is where schema 11 and
 * schema 18+ converge, where `cards.due`'s two meanings become one discriminated union, and
 * where `cards.ivl`'s sign convention becomes a plain duration. Everything downstream reads
 * the IR, so a quirk that survives to here survives into the database.
 *
 * Cross-references inside the IR use Anki's own numeric ids, which are unique within a
 * collection. Our UUIDv7 keys are minted by the database layer; `anki_id` columns keep these
 * for export fidelity (`docs/schema.md`).
 */

/** `notetypes` / the `models` JSON blob. */
export interface IrNoteType {
	ankiId: number;
	name: string;
	css: string;
	/** Cloze note types generate one card per `{{c<n>::…}}` ordinal, not one per template. */
	isCloze: boolean;
	sortFieldIdx: number;
	fields: IrField[];
	templates: IrTemplate[];
}

export interface IrField {
	ordinal: number;
	name: string;
	font: string;
	size: number;
	isRtl: boolean;
	sticky: boolean;
}

export interface IrTemplate {
	ordinal: number;
	name: string;
	qfmt: string;
	afmt: string;
	browserQfmt: string | null;
	browserAfmt: string | null;
}

export interface IrDeck {
	ankiId: number;
	/** Components joined by `::`, normalised from schema 18's `\x1f` separator. */
	name: string;
	description: string;
	/**
	 * Filtered ("dynamic") decks are a temporary study view, not a home for cards. Their cards
	 * carry a non-zero `originalDeckAnkiId` and belong to *that* deck; we have no equivalent
	 * concept, so the importer drops filtered decks and files their cards by origin.
	 */
	isFiltered: boolean;
}

export interface IrNote {
	ankiId: number;
	/** Anki's stable per-note id, and the idempotency key for re-import (CLAUDE.md §2.2). */
	guid: string;
	noteTypeAnkiId: number;
	/**
	 * The deck this note belongs to.
	 *
	 * Anki scopes decks to *cards*, so one note's cards can legitimately sit in several decks.
	 * Our schema scopes them to notes — `notes.deck_id` with `UNIQUE (deck_id, guid)` — because
	 * that unique index is the whole idempotency guarantee (CLAUDE.md §2.2), so the reader has to
	 * pick one.
	 *
	 * **Policy: the resolved home deck of the note's lowest-numbered card.** The rule has to be
	 * stable across re-imports of the same file or the upsert stops being a no-op and starts
	 * moving notes between decks, so it cannot depend on row iteration order, insertion order, or
	 * which deck happens to hold the most cards (ties). Anki card ids are creation timestamps and
	 * never change, which makes "lowest card id" both deterministic and the closest thing the
	 * collection has to "where this note was first filed".
	 *
	 * `null` only for a note with no cards at all — an orphan the database layer has nowhere to
	 * put and should drop.
	 */
	primaryDeckAnkiId: number | null;
	/** Field values in note-type field order, split from the `\x1f`-joined `flds` column. */
	fields: string[];
	tags: string[];
	/** Anki's truncated SHA-1 of the first field. Kept for export fidelity only. */
	checksum: number;
	createdAt: Date;
	modifiedAt: Date;
}

/**
 * `cards.due` means two unrelated things depending on the card's queue, and reading it wrong
 * produces a schedule that looks entirely plausible. Hence a union rather than a number:
 * downstream code cannot use a position as if it were a date without saying so.
 */
export type IrCardDue =
	/** New card: a *queue position*, not a time. Ordering only — it has no calendar meaning. */
	| { kind: 'position'; position: number }
	/** Intraday learning / preview: a wall-clock instant, stored as epoch seconds. */
	| { kind: 'timestamp'; at: Date }
	/** Review / day-learning: whole days since `col.crt`, resolved against `crt` into `at`. */
	| { kind: 'day'; dayOffset: number; at: Date };

/** How the card is held out of the queue. Orthogonal to `state` — Anki keeps it in `queue`. */
export type IrCardHold = 'none' | 'suspended' | 'buried-scheduler' | 'buried-user';

/** Anki's `cards.type`, which lines up with `ts-fsrs` `State`. */
export type IrCardState = 'new' | 'learning' | 'review' | 'relearning';

/** FSRS memory state from `cards.data`. Absent on cards FSRS never scheduled. */
export interface IrFsrsState {
	stability: number;
	difficulty: number;
	desiredRetention: number | null;
}

export interface IrCard {
	ankiId: number;
	noteAnkiId: number;
	/** Home deck: `odid` when the card is sitting in a filtered deck, otherwise `did`. */
	deckAnkiId: number;
	/** Set only while the card is in a filtered deck; the deck it is temporarily shown in. */
	filteredDeckAnkiId: number | null;
	/** Template ordinal, or cloze ordinal for cloze note types. */
	ordinal: number;
	state: IrCardState;
	hold: IrCardHold;
	due: IrCardDue;
	/**
	 * Current interval as a duration in seconds — the normalised form of `cards.ivl`'s
	 * days-or-negative-seconds encoding. Zero for cards with no interval yet.
	 */
	intervalSeconds: number;
	reps: number;
	lapses: number;
	/**
	 * SM-2 ease × 1000. Preserved for export fidelity and **never mapped to FSRS difficulty** —
	 * they are unrelated quantities (`docs/apkg-format.md`).
	 */
	sm2Factor: number;
	/** New-card position preserved in `cards.data` after the card leaves the new queue. */
	originalPosition: number | null;
	fsrs: IrFsrsState | null;
	flag: number;
	modifiedAt: Date;
}

/** Anki's `revlog.type`, matching `review_log.review_kind`. */
export type IrReviewKind = 'learn' | 'review' | 'relearn' | 'filtered' | 'manual';

export interface IrReview {
	/** `revlog.id` — epoch milliseconds, and unique per collection. */
	ankiId: number;
	cardAnkiId: number;
	reviewedAt: Date;
	/**
	 * Anki's `ease`: 1–4 for a graded review. `manual`/reschedule rows carry 0, which is not a
	 * rating; `review_log.rating` has a 1–4 check constraint, so those rows are the database
	 * layer's to drop or translate.
	 */
	ease: number;
	/** Interval after this review, normalised out of the days-or-negative-seconds encoding. */
	intervalSeconds: number;
	/** Interval before this review, same normalisation. */
	lastIntervalSeconds: number;
	sm2Factor: number;
	/** Time the answer took. Anki caps this at its per-deck maximum. */
	durationMs: number;
	kind: IrReviewKind;
}

/** One media file, keyed by the content hash `media_blobs` uses (`docs/schema.md`). */
export interface IrMediaBlob {
	/** Lowercase hex SHA-256. Also the storage path of the bytes. */
	sha256: string;
	sizeBytes: number;
	mime: string;
	bytes: Uint8Array;
}

/** The `media_refs` row: what a filename inside a note field resolves to. */
export interface IrMediaRef {
	/** NFC-normalised, as Anki stores it. */
	filename: string;
	sha256: string;
}

/**
 * Media is **package-scoped, not deck-scoped**, and deliberately so.
 *
 * `media_refs` is keyed `(deck_id, filename)`, which raises the same question `IrNote` answers
 * for notes — except that here there is nothing to disambiguate. A package carries one flat
 * media directory shared by every deck in it, and a filename's bytes do not vary by deck, so
 * there is no ordering-dependent choice to make and no stability risk. The database layer
 * writes one `media_refs` row per (deck the import created × filename), all pointing at the
 * same blob; `media_blobs` deduplicates globally by hash, so the repetition costs a row, not a
 * copy of the file.
 */
export interface IrMedia {
	/** Deduplicated by hash — two identical images in one package are one blob. */
	blobs: IrMediaBlob[];
	refs: IrMediaRef[];
}

/** Something survivable the reader had to decide about. Never thrown — the import continues. */
export interface IrWarning {
	code: 'media-duplicate-filename' | 'media-missing-member' | 'note-without-cards';
	message: string;
}

export interface IrCollection {
	/** `col.ver`: 11 and below is the JSON-blob layout, 18 and above the real-table layout. */
	schemaVersion: number;
	/**
	 * `col.crt` as an instant. The anchor every day-offset `due` is measured from, and the
	 * reason a review card's due date cannot be computed from the card row alone.
	 */
	createdAt: Date;
	noteTypes: IrNoteType[];
	decks: IrDeck[];
	notes: IrNote[];
	cards: IrCard[];
	reviews: IrReview[];
	media: IrMedia;
	/**
	 * Non-fatal oddities found while reading. A package with a colliding media filename or a
	 * cardless note is still worth importing; the user deserves to be told, not blocked.
	 */
	warnings: IrWarning[];
}

File: internal/apkg/ir.go
package apkg

import (
	"sort"
	"strconv"
	"strings"
	"time"
)

// IrCollection is one package, fully normalised. Both schema readers converge on this before any
// database code runs (architecture.md §4).
type IrCollection struct {
	Crt           time.Time // col.crt, UTC. An opaque anchor: used verbatim, nothing added.
	SchemaVersion int       // 11 or 18, from table presence -- never from col.ver
	NoteTypes     []IrNoteType
	Decks         []IrDeck
	Notes         []IrNote
	Cards         []IrCard
	Reviews       []IrReview
	Media         []IrMedia
	Warnings      []string // non-fatal findings, surfaced to the importing user
}

type IrNoteType struct {
	AnkiID       int64
	Name         string
	CSS          string
	IsCloze      bool
	SortFieldIdx int32
	Fields       []IrField    // sorted by Ordinal; ord is authoritative, array order is not
	Templates    []IrTemplate // sorted by Ordinal, same rule
}

type IrField struct {
	Ordinal int32
	Name    string
	Font    string
	Size    int32
	IsRTL   bool
	Sticky  bool
}

type IrTemplate struct {
	Ordinal     int32
	Name        string
	Qfmt        string
	Afmt        string
	BrowserQfmt string
	BrowserAfmt string
}

type IrDeck struct {
	AnkiID      int64
	Name        string // full path, "::"-separated in BOTH schemas -- schema 18's \x1f is normalised
	Description string
	IsFiltered  bool // dyn != 0. We have no filtered decks; cards are filed under their home deck
}

type IrNote struct {
	AnkiID         int64
	Guid           string
	NoteTypeAnkiID int64
	Fields         []string  // split on \x1f, indexed by IrField.Ordinal
	Tags           []string  // space-surrounded source string, empties dropped
	Checksum       int64     // notes.csum
	Created        time.Time // notes.id, epoch MILLISECONDS
	Modified       time.Time // notes.mod, epoch SECONDS
	// HomeDeckAnkiID is the resolved home deck of this note's LOWEST-numbered card
	// (architecture.md §20). It fills notes.deck_id: where the note was first filed, where the
	// notes list shows it, and the default for cards generated later. It is NOT what any card's
	// deck_id comes from -- that is IrCard.DeckAnkiID, per card.
	HomeDeckAnkiID int64
}

type IrDueKind uint8

const (
	DueNone     IrDueKind = iota // no meaningful due value
	DuePosition                  // new-card queue position; no calendar meaning
	DueAt                        // an absolute instant
)

// IrDue is cards.due after the three-way disambiguation in apkg-format.md, resolved from odue
// when the card sits in a filtered deck. The discriminator is `queue`, not `type`.
type IrDue struct {
	Kind     IrDueKind
	Position int32
	At       time.Time // UTC; valid only when Kind == DueAt
}

type IrCard struct {
	AnkiID     int64
	NoteAnkiID int64
	// DeckAnkiID is the card's real HOME deck: odid when odid != 0, else did. This is what
	// cards.deck_id is filed from (architecture.md §20's resolution -- the whole point of #58's
	// second bullet).
	DeckAnkiID         int64
	FilteredDeckAnkiID int64 // did when odid != 0, else 0. Carried for export fidelity only
	Ordinal            int32 // template ordinal, or cloze ordinal on a cloze note type
	Type               int32 // Anki cards.type, verbatim
	Queue              int32 // Anki cards.queue, verbatim
	Due                IrDue
	IntervalSeconds    int64 // cards.ivl normalised: days*86400 when positive, |ivl| when negative
	Factor             int32 // SM-2 ease x1000. NEVER mapped to FSRS difficulty (apkg-format.md)
	Reps               int32
	Lapses             int32
	Flag               int16 // cards.flags & 0x7
	Suspended          bool  // queue == -1
	Buried             bool  // queue == -2 or -3
	FSRS               *IrFSRSState // nil when cards.data carries no usable FSRS state
}

type IrFSRSState struct {
	Stability        float64
	Difficulty       float64
	DesiredRetention float64 // 0 when absent
	Position         int32   // cards.data "pos", the preserved new-card position; 0 when absent
}

type IrReview struct {
	AnkiID              int64     // revlog.id, epoch MILLISECONDS -- also the review instant
	CardAnkiID          int64
	ReviewedAt          time.Time // UTC, from AnkiID
	Rating              int16     // revlog.ease, 1..4. ease == 0 is a manual reschedule, dropped
	IntervalSeconds     int64     // revlog.ivl, same sign convention as cards.ivl
	LastIntervalSeconds int64     // revlog.lastIvl, same convention
	Factor              int32
	DurationMs          int32 // revlog.time
	Kind                int16 // revlog.type: 0 learning, 1 review, 2 relearning, 3 cram, 4 manual
}

type IrMedia struct {
	Index     string // the zip member name -- the numeric index the media index keyed it by
	Filename  string // NFC-normalised
	SHA256    string // lowercase hex
	SizeBytes int64
	Data      []byte
}

// intervalSeconds normalises Anki's dual-unit interval: days when positive, seconds when
// negative (apkg-format.md). Applies to cards.ivl, revlog.ivl and revlog.lastIvl alike -- the IR
// carries seconds throughout so the distinction cannot reappear downstream.
func intervalSeconds(ivl int64) int64 {
	if ivl < 0 {
		return -ivl
	}
	return ivl * 86400
}

// splitFields splits notes.flds on \x1f (unit separator).
func splitFields(flds string) []string {
	return strings.Split(flds, "\x1f")
}

// splitTags splits notes.tags, which is space-separated AND space-surrounded, dropping empties.
func splitTags(tags string) []string {
	fields := strings.Fields(tags)
	out := make([]string, 0, len(fields))
	out = append(out, fields...)
	return out
}

// normaliseDeckName converts schema 18's \x1f hierarchy separator to schema 11's "::".
// Idempotent, so it is safe to call on both readers' output.
func normaliseDeckName(name string) string {
	return strings.ReplaceAll(name, "\x1f", "::")
}

// resolveDue applies apkg-format.md's queue/odid/odue table.
//
// 1. If odid != 0, the value to interpret is odue, not due. Otherwise it is due.
// 2. Switch on queue, not type:
//   - 0 (new) -> DuePosition.
//   - 1 (learning), 4 (preview) -> DueAt, epoch seconds.
//   - 2 (review), 3 (day-learning) -> DueAt, days since crt. users.day_start_hour is not applied.
//   - negative (-1, -2, -3, held) -> the discriminating queue is gone. If type == 0, DuePosition.
//     Else if v >= epochSecondsThreshold, epoch seconds; else days since crt.
//   - anything else -> DueNone.
func resolveDue(queue, typ int32, due, odue, odid int64, crt time.Time) IrDue {
	v := due
	if odid != 0 {
		v = odue
	}
	switch queue {
	case ankiQueueNew:
		return IrDue{Kind: DuePosition, Position: int32(v)}
	case ankiQueueLearning, ankiQueuePreview:
		return IrDue{Kind: DueAt, At: time.Unix(v, 0).UTC()}
	case ankiQueueReview, ankiQueueDayLearning:
		return IrDue{Kind: DueAt, At: crt.Add(time.Duration(v) * 24 * time.Hour)}
	case ankiQueueSuspended, ankiQueueSchedBuried, ankiQueueUserBuried:
		if typ == ankiTypeNew {
			return IrDue{Kind: DuePosition, Position: int32(v)}
		}
		if v >= epochSecondsThreshold {
			return IrDue{Kind: DueAt, At: time.Unix(v, 0).UTC()}
		}
		return IrDue{Kind: DueAt, At: crt.Add(time.Duration(v) * 24 * time.Hour)}
	default:
		return IrDue{Kind: DueNone}
	}
}

// resolveHomeDecks fills every IrNote.HomeDeckAnkiID from the note's lowest-AnkiID card's
// DeckAnkiID (architecture.md §20). Notes with no cards keep 0 and are reported as a warning.
func resolveHomeDecks(notes []IrNote, cards []IrCard) []string {
	lowestCard := make(map[int64]IrCard, len(notes))
	for _, c := range cards {
		existing, ok := lowestCard[c.NoteAnkiID]
		if !ok || c.AnkiID < existing.AnkiID {
			lowestCard[c.NoteAnkiID] = c
		}
	}
	var warnings []string
	for i := range notes {
		c, ok := lowestCard[notes[i].AnkiID]
		if !ok {
			warnings = append(warnings, "note "+guidOrAnki(notes[i])+" has no cards; skipped")
			continue
		}
		notes[i].HomeDeckAnkiID = c.DeckAnkiID
	}
	return warnings
}

func guidOrAnki(n IrNote) string {
	if n.Guid != "" {
		return n.Guid
	}
	return "anki_id " + strconv.FormatInt(n.AnkiID, 10)
}

// sortFieldsByOrdinal sorts in place by Ordinal -- ord is authoritative, array order is not.
func sortFieldsByOrdinal(fields []IrField) {
	sort.Slice(fields, func(i, j int) bool { return fields[i].Ordinal < fields[j].Ordinal })
}

// sortTemplatesByOrdinal sorts in place by Ordinal -- ord is authoritative, array order is not.
func sortTemplatesByOrdinal(templates []IrTemplate) {
	sort.Slice(templates, func(i, j int) bool { return templates[i].Ordinal < templates[j].Ordinal })
}
